tutorbackup/templates/backup/build/backup/restore_services.py

```diff
@@ -0,0 +1,189 @@
+        elif os.path.isdir(item_path):
+            total_size += get_size(item_path)
+    return total_size
+
+
+def restore_mysql():
+    user = ENV['MYSQL_ROOT_USERNAME']
+    password = ENV['MYSQL_ROOT_PASSWORD']
+    host = ENV['MYSQL_HOST']
+    port = ENV['MYSQL_PORT']
+    dump_file = MYSQL_DUMPFILE
+
+    logger.info(f"Restoring MySQL databases on {host}:{port} from {dump_file}")
+    cmd = ("mysql "
+           f"--host={host} --port={port} "
+           f"--user={user} --password={password}")
+    with open(dump_file, 'rb') as dump:
+        check_call(cmd,
+                   shell=True,
+                   stdin=dump,
+                   stdout=sys.stdout,
+                   stderr=sys.stderr)
+
+    logger.info("MySQL restored.")
+
+
+def restore_mongodb():
+    host = ENV['MONGODB_HOST']
+    port = ENV['MONGODB_PORT']
+    dump_dir = MONGODB_DUMPDIR
+
+    logger.info(
+        f"Restoring MongoDB databases on {host}:{port} from {dump_dir}")
+    cmd = ("mongorestore "
+           "--stopOnError --drop "
+           f"--host={host} --port={port} "
+           f"{dump_dir}"
+           )
+    try:
+        cmd += (f" --username={ENV['MONGODB_USERNAME']} "
+                f"--password={ENV['MONGODB_PASSWORD']}")
+    except KeyError:
+        pass
+
+    check_call(cmd,
+               shell=True,
+               stdout=sys.stdout,
+               stderr=sys.stderr)
+
+    logger.info("MongoDB restored.")
+
+
+def restore_caddy():
+    dump_dir = CADDY_DUMPDIR
+    caddy_dir = 'caddy'
+    logger.info(f"Copying Caddy data from {dump_dir}")
+    shutil.copytree(dump_dir, caddy_dir)
+
+    total_size = get_size(caddy_dir)
+    logger.info(f"Complete. {caddy_dir} total size {total_size} bytes.")
+
+
+def extract():
+    tar_file = TARFILE
+    out_dir = DUMP_DIRECTORY
+
+    logger.info(f"Extracting archive {tar_file} to {out_dir}")
+    with tarfile.open(tar_file, "r:xz") as tar:
+        tar.extractall()
+
+    size = get_size(out_dir)
+    logger.info(f"Complete. {out_dir} is {size} bytes.")
+
+
+def download_from_s3(version_id=None):
+    from s3_client import S3_CLIENT, IntegrityError
+
+    bucket = ENV['S3_BUCKET_NAME']
+    file_name = TARFILE
+
+    logger.info(f"Downloading {file_name} from S3 bucket {bucket}")
+    try:
+        S3_CLIENT.download_file(
+            bucket,
+            os.path.basename(file_name),
+            file_name,
+            ExtraArgs={'VersionId': version_id} if version_id else None,
+        )
+
+        logger.info("Checking downloaded file's integrity ...")
+        obj_metadata = S3_CLIENT.head_object(
+            Bucket=bucket,
+            Key=os.path.basename(file_name),
+        )
+
+        received_size = obj_metadata['ContentLength']
+        size = os.path.getsize(file_name)
+        size_correct = received_size == size
+
+        received_version_id = obj_metadata['VersionId']
+        if version_id:
+            version_id_correct = version_id == received_version_id
```

Comment:
This is perfectly correct syntax, but I would tend to write it as

```python
version_id_correct = (version_id == received_version_id)
```

But then, some people see *that* and intuitively assume that `version_id_correct` is a tuple, not a boolean. :) Your choice.

Reply:
I also used to write it like that and got feedback that those parentheses are redundant. I agree that with parentheses, the right side of the assignment is visually clearer. So I'll go back to my old ways :)

Reply:
I will always go for redundant parentheses if they improve legibility. Though, like I said, in this case it can be argued that they do (because they make the `=` vs. `==` distinction clearer) and also that they don't (because they might make the variable look like a tuple).